### PR TITLE
Improved AskPositiveHandler and AskNegativeHandler

### LINKED ISSUE #14993: assumptions fail on simple Q.Positive expression
sympy1.2: ask(Q.positive(-x-1), Q.positive(x)) returns None
This used to work (return False) on sympy 1.0

### PR DESCRIPTION
**Pull request by ShubhamKJha**
#### References to other Issues or PRs

Fixes #14993 

#### Brief description of what is fixed or changed
Improved the code of `AskPositiveHandler` and `AskNegativeHandler`.
Earlier
```py
>>> from sympy import Q, ask
>>> from sympy.abc import x
>>> ask(Q.positive(-x-1), Q.positive(x))
>>> ask(Q.negative(-x+1), Q.negative(x))
>>>
```
But now
```py
>>> ask(Q.positive(-x-1), Q.positive(x))
False
>>> ask(Q.negative(-x+1), Q.negative(x))
False
>>>
```
#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by ShubhamKJha** [comment]
@Abdullahjavednesar If I understand the Code coverage correctly the tests are failing here because the control doesn't reach some parts while testing. But there are already some tests reaching that part (marked `@slow`). Does CodeCov not consider such tests? How should I approach it? I can remove the `@slow` decorator or create new faster tests specifically for those segments. Please suggest.

**Comment by asmeurer** [comment]
Codecov currently isn't including the slow tests, since coverage slows down the test runs on Travis, so they would slow down the slow test runs too much. If it is already covered by the slow tests you can ignore it.

File: sympy/assumptions/handlers/order.py [hunk 1]
```diff
@@ -272,16 +284,28 @@ def Add(expr, assumptions):
...
-            if nonneg < len(expr.args):
+        # Helper method to check positive or negative
+        # Returns True if positive, False if negative or None otherwise
+        def neg_pos(expr):
+            if ask(Q.negative(expr), assumptions):
+                return False
+            elif ask(Q.positive(expr), assumptions):
                 return True
+            elif ask(Q.nonnegative(expr) | Q.nonpositive(expr), assumptions):
+                return 0
```
Comment (asmeurer) [inline sympy/assumptions/handlers/order.py]: Why is this 0 instead of False?

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: `0` is provided when the `assumptions` is apparently `S.Zero`.

**Comment by asmeurer** [comment]
Anyway, if this fixes behavior that didn't work before, then new tests should be added for it.

File: sympy/assumptions/handlers/order.py (hunk 1 above)
Comment (asmeurer) [reply sympy/assumptions/handlers/order.py]: But the result is only used in a boolean context, so it's the same as False.

**Comment by asmeurer** [comment]
We should check the logic here carefully. Is it possible to reuse some of the old assumptions logic for this?

File: sympy/assumptions/handlers/order.py (hunk 1 above)
Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: I needed four results for the assumptions.
 - Positive
- Negative
- Zero
- Unknown

I used `True`, `False`, `None` so I put `0` there

**Comment by asmeurer** [comment]
Either way, the would be easier to follow if it used the fuzzy logic functions.

File: sympy/assumptions/handlers/order.py (hunk 1 above)
Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: > But the result is only used in a boolean context, so it's the same as False.

I have used `is` everywhere. Hence it should not create a problem. My strategy is to ignore if something which can take a form of `0` occurs and halt if `Unknown` occurs.

**Comment by ShubhamKJha** [comment]
Considering your suggestions I would give it some more thought.

File: sympy/assumptions/handlers/order.py (hunk 1 above)
Comment (asmeurer) [reply sympy/assumptions/handlers/order.py]: Is it intended that `0` will pass the `in (True, False)` check but not the `is (not val)` check? This is very opaque if it is. I would find a clearer way to write the algorithm.

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: I am currently writing a better algorithm. This already doesn't pass many cases I thought of.

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: Is it only with `imaginary` symbols that `Q.positive(), Q.negative(), Q.nonnegative() and Q.nonpositive()` all result in `False`? Or there are other such cases. I am trying to make it work for all cases.

**Comment by ShubhamKJha** [comment]
@asmeurer please take a look at this.

**Comment by ShubhamKJha** [comment]
@jksuom please review this.

**Comment by jksuom** [comment]
This looks pretty complicated to me. I am afraid it would be hard to maintain. I would start improving these handlers by adding new specialized code (for `Add`, `Mul`, `Pow`) to `AskNonNegativeHandler` and `AskNonPositiveHandler`, for example.

**Comment by ShubhamKJha** [comment]
@jksuom I don't understand.
I focused on only `Add` part where I got 15 special cases of interaction among two arguments( which can be either `positive`,`negative`,`non-negative`,`non-positive`, or `zero`). Based on these result is inferred. Most of these cases were left in the previous implementation.
Are you suggesting to improve other parts of the handler as well?

**Comment by jksuom** [comment]
I think that there are handlers where the basic cases of Add have not been implemented yet. That is where I would start.

**Comment by ShubhamKJha** [comment]
Ok, I would also work on them. But part of my GSoC proposal is to switch handlers with their Relational equivalents. As you can look in this code there were so many predicates dominating the end result. And I had to take each of them into concern. If they were converted to their Relational equivalents then so much logic might not have been needed. 
In any case, this PR is working and fixing the issue (with added capabilities). If the code looks good it can be merged I believe.

**Comment by ShubhamKJha** [comment]
> Is it possible to reuse some of the old assumptions logic for this?

I went through old assumptions in the core and improved the code for these handlers. @asmeurer @jksuom please review this PR. I have added some functions to handlers which were required for this PR. I will add more functions to handlers with time.

**Comment by ShubhamKJha** [comment]
Another inconsistency I found in old assumptions is:
```py
>>> from sympy import symbols
>>> p = symbols('p', zero=True)
>>> q = symbols('q', even=True, zero=False)
>>> p.is_zero
True
>>> q.is_even
True
>>> (p**q).is_positive   # 0**(even != 0) should be 0
>>>
```

**Comment by ShubhamKJha** [comment]
@jksuom @asmeurer please tell if the above seems right.

**Comment by ShubhamKJha** [comment]
@Abdullahjavednesar please remove the `author's turn` label, I need the code reviewed before adding further code.

File: sympy/assumptions/handlers/order.py [hunk 2]
```diff
@@ -56,43 +56,95 @@ def Basic(expr, assumptions):
...
+        if not args:
+            return False
+
+        inf = neg_inf = pos_inf = neg = npos = nneg = False
+
+        for arg in args:
+            n  = ask(Q.negative(arg), assumptions)
+            np = ask(Q.nonpositive(arg), assumptions)
+
+            infinite = ask(Q.infinite(arg), assumptions)
```
Comment (jksuom) [inline sympy/assumptions/handlers/order.py]: Would it not suffice to know the sign of `arg`, whether finite or infinite?

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: if any one argument is `+oo` it doesn't matter the other args are what sign. Similarly for `-oo` but if both `oo` and `-oo` happens to be present the result becomes ambiguous. Most of such ambiguities are handled.

Comment (jksuom) [reply sympy/assumptions/handlers/order.py]: What if one argument is positive?

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: I don't understand. All the arguments are checked. If positive and +oo are there then +oo would be resulted hence positive.

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: And with -oo and a positive argument the result would be -oo always.

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: We need to check what is overall effect of all the arguments is on the expression.

Comment (jksuom) [reply sympy/assumptions/handlers/order.py]: Would it not be possible to use AskNonNegativeHandler to take care of positive and +oo?

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: Yeah actually I can use the code is pretty much repetitive. Thanks for suggestion.

Comment (ShubhamKJha) [reply sympy/assumptions/handlers/order.py]: I will see this case but I intended to make AskNegativeHandler and AskPositiveHandler the base handlers from them I would derive AskNonNegativeHandler and others. In this way much repetitive code can be reduced.

**Comment by ShubhamKJha** [comment]
> Would it not be possible to use AskNonNegativeHandler to take care of positive and +oo?

@jksuom either AskPositiveHandler can be defined using AskNonNegativeHandler and AskZeroHandler or the other way round. I believe it is better to let AskPositive and AskNegative be the ones over which others should be defined. Please take a look if it can be merged.

**Comment by jksuom** [comment]
Can you give examples where this implementation of `AskNegativeHandler.Add` (with test of `Q.infinite`) will be needed?

**Comment by ShubhamKJha** [comment]
The following works now
```py
>>> from sympy import Q, ask
>>> from sympy.abc import x,y,z
>>> ask(Q.negative(x+y), Q.negative(x) & Q.positive(y) & Q.infinite(y))
False
>>>
```
This is how it works in old assumptions
```py
In [43]: a = Symbol('a', infinite=True, positive=True)

In [44]: b = Symbol('b', negative=True)

In [45]: (a+b).is_negative  # oo + anything == oo
Out[45]: False
```

**Comment by jksuom** [comment]
> ask(Q.negative(x+y), Q.negative(x) & Q.positive(y) & Q.infinite(y))

Is it clear that `x` could not be `-oo`?

**Comment by ShubhamKJha** [comment]
You are right. `x` can be `-oo` as well. I will update code. But it also means that `core` would need to be updated as well.

**Comment by ShubhamKJha** [comment]
Hi @jksuom please take a look at this.

File: sympy/assumptions/handlers/order.py [hunk 3]
```diff
@@ -56,43 +56,98 @@ def Basic(expr, assumptions):
...
+                elif nn:
+                    nneg = True
+                    continue
                 else:
-                    break
-        else:
-            if nonpos < len(expr.args):
-                return True
+                    return None  # Unknown symbol
+        if inf and pos_inf:
```
Comment (jksuom) [inline sympy/assumptions/handlers/order.py]: Is it necessary to have two flags, `inf` and `pos_inf`?

**Comment by jksuom** [comment]
I would preferably avoid using flags to control the code flow if possible. Would it be possible to do something like this: First scan the arguments for infinities. If there are none, then look at all arguments.

**Comment by ShubhamKJha** [comment]
The calls to `ask` are still costly. I tried to keep such calls minimum through the use of flags.

> First scan the arguments for infinities. If there are none, then look at all arguments.

There can be a large number of `arguments`. A scan would terminate when it is not possible to infer 'positive','negative',etc. The above process would still check infiniteness of all the args.

**Comment by jksuom** [comment]
> The above process would still check infiniteness of all the args

That is not necessary if two infinities of opposite signs are encountered.

**Comment by ShubhamKJha** [comment]
@jksuom I meant if there are hundreds of arguments and on the first argument no assumptions can be inferred, then the current approach will stop right there. But checking infinities as you asked, it should check all the remaining arguments.

**Comment by jksuom** [comment]
Considering that `ask` may be expensive, is it necessary to call it twice before checking if the argument is infinite?

**Comment by ShubhamKJha** [comment]
I need to know if the argument is positive or negative whatever infinity check returns(`True`,`False` or `None`).

**Comment by asmeurer** [comment]
I agree the code here is pretty complicated. This shows why the handlers design for the assumptions isn't so great. You have to add more and more logic to handle more and more cases, and the code gets harder to read. Contrast that with the satask way, where to add more cases you add more facts, which are just single line predicates which can be easily verified. The actual logic of how the deductions are made is handled by the solver. Only the actual assumptions facts are written down.  The exception is certain "metapredicates" on top of the predicates to handle advanced deductions (for instance, order properties). Representing things using inequalities would be even better for readability (and obviously it would be a boon for end user code as well). The only issue with it is the performance. 

For this PR, I see lots of branches. I would make sure that every new line of code is covered by tests (see the "diff" tab on codecov). Based on the current codecov, there are a lot of lines missing coverage.

**Comment by ShubhamKJha** [comment]
>  This shows why the handlers design for the assumptions isn't so great.

Hi @asmeurer I agree the design of handler mechanism is not good at all. The same applies to old assumptions from which it derives.
> Based on the current codecov, there are a lot of lines missing coverage.

I think codecov currently doesn't work on `@slow` tests hence the missing coverage.
> I agree the code here is pretty complicated

I have cleaned the code a little bit. Please take a look at it.

**Comment by ShubhamKJha** [comment]
@jksuom please take a look at it if it needs to be improved.

**Comment by ShubhamKJha** [comment]
Hi @jksuom please take a look at it.

**Comment by jksuom** [comment]
I would start by checking if there are infinities before looking into the sign of any argument. If one is found, then it would suffice to consider only the signs of infinite arguments. Otherwise, all signs have to be investigated.

**Comment by ShubhamKJha** [comment]
I have also considered the case of 'may-be infinites' (where `infinite` is `None`) and their interactions with `finite` and `infinite` values. This makes the separation between `infinite` and `finite` values difficult.The present code handles all of these cases. I agree the code could look complex but it is easy to understand and maintain.

**Comment by czgdp1807** [comment]
@ShubhamKJha The associated issue is still unresolved. Would you like to work on it? Thanks for the contributions.

**Comment by Arpan612** [comment]
@czgdp1807 @jksuom Hello. May I continue the work on this issue?

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/sympy/sympy/pull/19241

**Pull request closed without merging**